### PR TITLE
State: Sync sites-list to Redux state

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to sync sites data from the legacy store to Redux state in an attempt to facilitate migration toward removing `lib/sites-list` altogether.

There has been previous discussions (p4TIVU-3B9-p2) and attempts (#2167) to address the need for all sites within Redux state. While the data duplication resulting from the migratory sync is not ideal (memory usage, potential for lack of follow-through), I'd argue there's not been presented a compelling alternative.

__Implementation notes:__

A few additional minor changes were included:

- The site `meta` property was removed from its schema, so as to avoid persisting this. Because we are likely storing much more data (especially for users with many sites), storage quotas become a concern. The `meta` property should likely never be necessary to have in the context of Calypso.
- The `plan.expired` property was set as required, but this property is not always present for some sites (notably, VIP sites). See r135973-wpcom, cc @eoigal, @gwwar 

__Testing instructions:__

There are no visual changes in this pull request. It's easiest to verify expected behavior using the [Redux DevTools Chrome extension](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en). After loading the Calypso application, open the Chrome DevTools to the Redux tab, enter `SITES_RECEIVE` as a filter (or look for this action within the list). Note that the dispatched action includes an array of all sites received via the `lib/sites-list` request and that the affected state includes updating `state.sites.items` and `state.currentUser.capabilities` for the received sites.

cc @mtias 

Test live: https://calypso.live/?branch=add/sites-list-redux-sync

### CODE REVIEW

**Comment by aduth** [comment]
For context, desire for this grew from second caveat in #7593. We should update Jetpack current version (`site.options.jetpack_version`) in state for a site after updating the Jetpack plugin through the Calypso UI. Plugin updates are currently managed through the [Flux-based stores](https://github.com/Automattic/wp-calypso/blob/e031a1720d356ad7b915094e479ab03b56a6bc13/client/lib/plugins/actions.js#L217-L249). If we can sync sites, the thought is that maybe `sites-list` could listen for this action through the Flux Dispatcher object and update its internal state. We should seek to move forward with Redux-ifying plugins altogether, but I'm not aware of this being on anyone's radar for the immediate future and seems like a much larger undertaking.

File: client/state/sites/middleware.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Internal dependencies
+ */
+import sitesList from 'lib/sites-list';
+import { receiveSites } from 'state/sites/actions';
+
+export default ( { dispatch } ) => {
+	const sites = sitesList();
+	sites.on( 'change', () => {
+		dispatch( receiveSites( sites.get() ) );
```
Comment (mtias) [inline client/state/sites/middleware.js]: Could we have issues with this `sites.get()` not being pure? For instance, `getRawSite()` would be selecting an already computed site object, and the operations we do on `getSite` could trip over expecting some of these fields to be empty?

Comment (aduth) [inline client/state/sites/middleware.js]: >Could we have issues with this `sites.get()` not being pure?

This should be handled the same way as how we [receive sites into Redux from `sites-list` in the `my-sites` controller](https://github.com/Automattic/wp-calypso/blob/6bb2e62f614bd6d58828b27e8db87a36d4482062/client/my-sites/controller.js#L169-L171). Which is: Received sites will have properties [picked using "valid" site keys from the state schema](https://github.com/Automattic/wp-calypso/blob/6bb2e62f614bd6d58828b27e8db87a36d4482062/client/state/sites/reducer.js#L66), and computed properties merged into when selecting [using the `getSite` selector](https://github.com/Automattic/wp-calypso/blob/6bb2e62f614bd6d58828b27e8db87a36d4482062/client/state/sites/selectors.js#L42-L70).

**Comment by gwwar** [comment]
> The plan.expired property was set as required, but this property is not always present for some sites (notably, VIP sites)

Thanks for fixing that 😄

**Comment by gwwar** [comment]
👍  This works as expected. I'd be fine with merging this in, provided we can wrangle other devs to help with the sites-list port. We won't want this data duplication to live very long in the app.

**Comment by aduth** [comment]
Per our discussion about sites list refactoring, I'm going to mark this as punted 'til we can make more incremental progress on sites list consuming components.

**Pull request closed without merging**